cheatsheets/Password_Storage_Cheat_Sheet.md

```diff
@@ -1,567 +1,153 @@
-1. Password-generated session key - fraction of a second; 
-2. User credential - ~0.5 seconds; 
-3. Password-generated site (or other long-lived) key - potentially a second or more.
-
-## Ref7
-
-- [Argon2 detailed specifications can be found here](https://password-hashing.net/argon2-specs.pdf).
-
-## Ref8
-
-- [Painless Password Hash Upgrades](https://veggiespam.com/painless-password-hash-upgrades/).
+- Use the strongest algorithm available (SHA-512 > SHA-256 > SHA-1 > MD5).
+- Use a [pepper](#peppering).
+- Use a unique [salt](#salting) for each password, generated using a [cryptographically secure random number generator](Cryptographic_Storage_Cheat_Sheet.md#rule---use-cryptographically-secure-pseudo-random-number-generators-csprng).
+- Use a very large number of iterations of the algorithm (at least 10,000, and possibly significantly more depending on the speed of the hardware).
+
+It should be emphasised that these steps **are not as good as using a  modern hashing algorithm**, and that this approach should only be taken where no other options are available.
+
+## Upgrading Legacy Hashes
+
+For older applications that were built using less secure password hashing algorithms such as MD5 or SHA-1, there are a number of different methods that can be undertaken to upgrade the password hashes into a more secure algorithm.
+
+The simplest way to do this is to use the existing password hashes as inputs for a more secure algorithm. For example if the application originally stored passwords as `md5(password)`, this could be easily upgraded to `argon2(md5(password))`. Stacking the hashes in this manner avoids the need to known the original password, and can also solve potential issues related to [password length](#maximum-password-lengths).
+
+An alternative approach is to wait until the user enters their password (by authenticating on the application), and then replacing their old password hash with a new one using a modern algorithm. The downside of this is that old, weak hashes will remain in the database until the users next login, which could mean that they are never replaced if users are inactive. One way to solve this is to expire and delete the password hashes of users who have been inactive for a long period, and require them to reset their passwords to login again. This approach also requires more complicated authentication code, which needs to determine the type of hash stored, compare securely against it, and then overwrite it. There are a number of [articles](https://veggiespam.com/painless-password-hash-upgrades/) that discuss these issues and other potential approaches.
+
+## Custom Algorithms
+
+Writing custom cryptographic code such as a hashing algorithm is **really hard** and should **never be done** outside of an academic exercise. Any potential benefit that you might have from using an unknown or bespoke algorithm will be vastly overshadowed by the weaknesses that exist in it.
+
+**Do not do this.**
```

Comment:
Why remove the code snippets? I believe we should remove them if we have a way to replace them, unless we put them in another file 😄

Reply:
Most of the code blocks aren't very useful or relevant to the wider audience.

There was:

- A class for hashing passwords in Java (only relevant to a small subset of users, and also a few years old, so I don't know how correct or working it is with the latest Java)
- Some test code to demonstrate that Argon2 can hash long passwords (not useful)
- A script showing how to compile an old version of Argon2 with an old version of PHP (not useful, and outdated)
- A class for hashing passwords in PHP (only relevant to a small subset of users,  doesn't really add much above the default functions, and also relies on a `config.ini` file adding extra complexity)

For the vast majority of users, I don't think these code samples really added anything on value. Maintaining secure cryptographic libraries is **really** hard, so I'm not sure it's really something that we want to be taking on.

Reply:
Let's try linking to the official repos then?
Something like this, which tells the developer the bindings for every language that implements argon2: https://github.com/P-H-C/phc-winner-argon2#bindings

Reply:
I found different ones for bcrypt, which won't help us much, as we'll be back to maintaining them. I believe general recommendations can be taken for certain parameters that can be used? The algorithm is the same, the parameters are the same. We can focus on that, and let the developer either follow the best practices set (at the time that we set this guide), where we mention that it was *to this date*, the best practices.

Reply:
Linked to the bindings page in line 124.

Reply:
Trying to give sample code (or even links) for every language is going to become pretty unmaintainable - but in most cases the language docs should cover it. For example, with PHP if you google `PHP argon2`, then the first link you find is the official PHP docs which explain how to use Argon2, and the default work factors/etc that it has.